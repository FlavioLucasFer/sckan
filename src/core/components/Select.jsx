import React, { useEffect } from 'react';
import PropTypes from 'prop-types';

import Icon from 'core/components/Icon';

Select.propTypes = {
	title: PropTypes.string.isRequired,
	className: PropTypes.string,
	value: PropTypes.any,
	values: PropTypes.arrayOf(Object).isRequired,
	defaultOption: PropTypes.string,
	icon: PropTypes.string,
	options: PropTypes.object,
	valueAs: PropTypes.string,
	titleAs: PropTypes.string,
};

function Select(props) {
	const { 
		title, 
		className,
		values,
		defaultOption, 
		icon,
		options,
		valueAs,
		titleAs,
		...rest 
	} = props;

	useEffect(() => {
		const elems = document.querySelectorAll('select');
		M.FormSelect.init(elems, options);
	}, [values, props.value]);
	
	function renderOptions() {
		if (values.length <= 0)
			return <option value="" disabled selected>No data found!</option>

		return values.map((e, i) => (
			<option key={i} value={valueAs ? e[valueAs] : e.value}>
				{titleAs ? e[titleAs] : e.title}
			</option>
		));
	}

	function renderDefaultOption() {
		if (!defaultOption || values.length === 0)
			return null;

		return (
			<option value="" disabled selected>{defaultOption}</option>
		);
	}

	function renderIcon() {
		if (!icon)
			return null;

		return (
			<Icon className="prefix"
				color="grey-text">
				{icon}
			</Icon>
		);
	}

	return (
		<div className={`input-field ${className}`}>
			{renderIcon()}
			<select {...rest}>
				{renderDefaultOption()}
				{renderOptions()}
			</select>
			<label>{title}</label>
		</div>
	);
}

export default Select;
