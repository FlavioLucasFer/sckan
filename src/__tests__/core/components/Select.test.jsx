import React from 'react';
import { shallow } from 'enzyme';

import Select from 'core/components/Select';

describe('Testing Select core component', () => {
	it('should render correctly', () => {
		let values = [
			{ title: 'option1', value: 'opt1' },
			{ title: 'option2', value: 'opt2' },
			{ title: 'option3', value: 'opt3' },
		];
		
		const wrapper = shallow(
			<Select title="select test"
				values={values} />
		);

		expect(wrapper.debug()).toMatchSnapshot();
		
		values = [
			{ title: 'option4', value: 'opt4' },
			{ title: 'option5', value: 'opt5' },
		];

		wrapper.setProps({
			title: 'testing',
			values,
			className: 'col s12',
			defaultOption: 'select a option',
			icon: 'business',
			options: {
				classes: 'teal',
			},
		});
		
		expect(wrapper.debug()).toMatchSnapshot();
	});
});