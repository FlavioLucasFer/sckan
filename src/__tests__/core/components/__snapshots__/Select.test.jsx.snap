// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Select core component should render correctly 1`] = `
"<div className=\\"input-field undefined\\">
  <select>
    <option value=\\"opt1\\">
      option1
    </option>
    <option value=\\"opt2\\">
      option2
    </option>
    <option value=\\"opt3\\">
      option3
    </option>
  </select>
  <label>
    select test
  </label>
</div>"
`;

exports[`Testing Select core component should render correctly 2`] = `
"<div className=\\"input-field col s12\\">
  <Icon className=\\"prefix\\" color=\\"grey-text\\">
    business
  </Icon>
  <select>
    <option value=\\"\\" disabled={true} selected={true}>
      select a option
    </option>
    <option value=\\"opt4\\">
      option4
    </option>
    <option value=\\"opt5\\">
      option5
    </option>
  </select>
  <label>
    testing
  </label>
</div>"
`;
